// parse-server/lib/graphql/types/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.type = type;
exports.inputType = inputType;
exports.queryType = queryType;
Object.defineProperty(exports, "ACL", {
  enumerable: true,
  get: function () {
    return _ACL.ACL;
  }
});
Object.defineProperty(exports, "ACLInput", {
  enumerable: true,
  get: function () {
    return _ACL.ACLInput;
  }
});
Object.defineProperty(exports, "GeoPoint", {
  enumerable: true,
  get: function () {
    return _GeoPoint.GeoPoint;
  }
});
Object.defineProperty(exports, "GeoPointInput", {
  enumerable: true,
  get: function () {
    return _GeoPoint.GeoPointInput;
  }
});
Object.defineProperty(exports, "File", {
  enumerable: true,
  get: function () {
    return _File.File;
  }
});
Object.defineProperty(exports, "FileInput", {
  enumerable: true,
  get: function () {
    return _File.FileInput;
  }
});
Object.defineProperty(exports, "Date", {
  enumerable: true,
  get: function () {
    return _Date.Date;
  }
});
Object.defineProperty(exports, "Pointer", {
  enumerable: true,
  get: function () {
    return _Pointer.Pointer;
  }
});
Object.defineProperty(exports, "JSONObject", {
  enumerable: true,
  get: function () {
    return _JSONObject.JSONObject;
  }
});
Object.defineProperty(exports, "PageInfo", {
  enumerable: true,
  get: function () {
    return _PageInfo.PageInfo;
  }
});
exports.BooleanQuery = void 0;

var _graphql = require("graphql");

var _ACL = require("./ACL");

var _GeoPoint = require("./GeoPoint");

var _File = require("./File");

var _Date = require("./Date");

var _Pointer = require("./Pointer");

var _JSONObject = require("./JSONObject");

var _StringQuery = require("./StringQuery");

var _NumberQuery = require("./NumberQuery");

var _NumberInput = require("./NumberInput");

var _PageInfo = require("./PageInfo");

var _Array = require("./Array");

var _BaseQuery = require("./BaseQuery");

const types = {
  String: _graphql.GraphQLString,
  Number: _graphql.GraphQLFloat,
  Boolean: _graphql.GraphQLBoolean,
  GeoPoint: _GeoPoint.GeoPoint,
  File: _File.File,
  ACL: _ACL.ACL,
  Date: _Date.Date,
  Pointer: _Pointer.Pointer,
  Object: _JSONObject.JSONObject,
  Array: new _graphql.GraphQLList(_JSONObject.JSONObject)
};
const BooleanQuery = new _graphql.GraphQLInputObjectType({
  name: 'BooleanQuery',
  fields: (0, _BaseQuery.BaseQuery)(_graphql.GraphQLBoolean)
});
exports.BooleanQuery = BooleanQuery;

function type({
  type
}) {
  return types[type];
}

function inputType(field) {
  const {
    type
  } = field;

  if (type == 'String') {
    return _graphql.GraphQLString;
  }

  if (type == 'Number') {
    return _NumberInput.NumberInput;
  }

  if (type == 'Boolean') {
    return _graphql.GraphQLBoolean;
  }

  if (type == 'GeoPoint') {
    return _GeoPoint.GeoPointInput;
  }

  if (type == 'File') {
    return _File.FileInput;
  } else if (type == 'ACL') {
    return _ACL.ACLInput;
  } else if (type == 'Date') {
    return _Date.Date;
  } else if (type == 'Pointer') {
    return (0, _Pointer.PointerInput)(field);
  } else if (type === 'Array') {
    return new _graphql.GraphQLList(_JSONObject.JSONObject);
  } else if (type === 'Object') {
    return _JSONObject.JSONObject;
  }
}

function queryType(field) {
  const {
    type
  } = field;

  if (type == 'String') {
    return _StringQuery.StringQuery;
  }

  if (type == 'Number') {
    return _NumberQuery.NumberQuery;
  }

  if (type == 'Boolean') {
    return BooleanQuery;
  }

  if (type == 'GeoPoint') {
    return _GeoPoint.GeoPointQuery;
  }

  if (type == 'File') {
    // Cannot query on files
    return;
  } else if (type == 'ACL') {
    // cannot query on ACL!
    return;
  } else if (type == 'Date') {
    return _Date.DateQuery;
  } else if (type == 'Pointer') {
    return (0, _Pointer.PointerQuery)(field);
  } else if (type == 'Array') {
    return _Array.ArrayQuery;
  }
}
