// parse-server/lib/graphql/ParseQuery.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AtomicOps = exports.ParseQuery = void 0;

var _graphql = require("graphql");

const FloatKeyValIT = new _graphql.GraphQLInputObjectType({
  name: 'FloatConstraint',
  fields: {
    key: {
      type: _graphql.GraphQLString
    },
    value: {
      type: _graphql.GraphQLFloat
    }
  }
});
const StringKeyValIT = new _graphql.GraphQLInputObjectType({
  name: 'StringConstraint',
  fields: {
    key: {
      type: _graphql.GraphQLString
    },
    value: {
      type: _graphql.GraphQLString
    }
  }
});
const ParseQuery = {
  whereLessThan: {
    type: FloatKeyValIT,
    args: FloatKeyValIT.fields,
    description: ''
  },
  whereGreaterThan: {
    type: FloatKeyValIT,
    args: FloatKeyValIT.fields
  },
  whereLessThanOrEqualTo: {
    type: FloatKeyValIT,
    args: FloatKeyValIT.fields
  },
  whereGreaterThanOrEqualTo: {
    type: FloatKeyValIT,
    args: FloatKeyValIT.fields
  },
  whereMatches: {
    type: StringKeyValIT,
    args: StringKeyValIT.fields
  },
  whereExists: {
    type: _graphql.GraphQLString,
    args: {
      key: _graphql.GraphQLString
    }
  },
  whereDoesNotExist: {
    type: _graphql.GraphQLString,
    args: {
      key: _graphql.GraphQLString
    }
  },
  whereStartsWith: {
    type: StringKeyValIT,
    args: StringKeyValIT.fields
  },
  whereEndsWith: {
    type: StringKeyValIT,
    args: StringKeyValIT.fields
  }
};
/* eslint-disable */

exports.ParseQuery = ParseQuery;
const AtomicOps = {
  incrementKey: {
    type: FloatKeyValIT,
    args: FloatKeyValIT.fields
  },
  unsetKey: {
    type: _graphql.GraphQLString,
    args: {
      key: _graphql.GraphQLString
    }
  }
};
exports.AtomicOps = AtomicOps;
