// parse-server/lib/graphql/Schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GraphQLParseSchema = void 0;

var _graphql = require("graphql");

var _Config = _interopRequireDefault(require("../Config"));

var _ParseClass = _interopRequireDefault(require("./schemas/ParseClass"));

var _UserAuth = _interopRequireDefault(require("./schemas/UserAuth"));

var _Node = _interopRequireDefault(require("./schemas/Node"));

var _Functions = _interopRequireDefault(require("./schemas/Functions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class GraphQLParseSchema {
  constructor(applicationId) {
    this.applicationId = applicationId;
  }

  static async loadSchemaFromDatabase(applicationId) {
    const schema = await _Config.default.get(applicationId).database.loadSchema();
    const allClasses = await schema.getAllClasses(true);
    const classNames = [];
    const fullSchema = allClasses.reduce((memo, classDef) => {
      memo[classDef.className] = classDef;
      classNames.push(classDef.className);
      return memo;
    }, {});
    fullSchema.__classNames = classNames;
    return Object.freeze(fullSchema);
  }

  async load() {
    this.schema = await GraphQLParseSchema.loadSchemaFromDatabase(this.applicationId);
    const graphQLSchema = new _graphql.GraphQLSchema({
      query: this.Query(),
      mutation: this.Mutation()
    });
    return {
      schema: graphQLSchema,
      rootValue: this.Root()
    };
  }

  Query() {
    return new _graphql.GraphQLObjectType({
      name: 'Query',
      description: `The query root of you Parse Server's graphQL interface`,
      fields: () => {
        const fields = {};
        Object.assign(fields, _Node.default.Query(this.schema));
        Object.assign(fields, _UserAuth.default.Query(this.schema));
        Object.assign(fields, _ParseClass.default.Query(this.schema));
        return fields;
      }
    });
  }

  Mutation() {
    const fields = {};
    Object.assign(fields, _UserAuth.default.Mutation(this.schema));
    Object.assign(fields, _ParseClass.default.Mutation(this.schema));
    Object.assign(fields, _Functions.default.Mutation(this.schema));
    return new _graphql.GraphQLObjectType({
      name: 'Mutation',
      fields
    });
  }

  Root() {
    return this.schema.__classNames.reduce((memo, className) => {
      memo[className] = {};
      return memo;
    }, {});
  }

}

exports.GraphQLParseSchema = GraphQLParseSchema;
