// parse-server/lib/graphql/UserAuth.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getUserAuthMutationFields = getUserAuthMutationFields;

var _graphql = require('graphql');

var _execute = require('./execute');

var _UserAuthentication = require('../Controllers/UserAuthentication');

var _ParseClass = require('./ParseClass');

const LoginCompletePayload = new _graphql.GraphQLObjectType({
  name: 'LoginCompletePayload',
  fields: () => {
    const { parseClass } = (0, _ParseClass.loadClass)('_User');
    const fields = parseClass.graphQLConfig().fields;
    return Object.assign({}, fields(), {
      sessionToken: { type: (0, _graphql.GraphQLNonNull)(_graphql.GraphQLString) }
    });
  }
});

const LoginInput = new _graphql.GraphQLInputObjectType({
  name: 'LoginInput',
  fields: {
    email: { type: _graphql.GraphQLString, description: 'the email of the user. Either email or username should be provided' },
    username: { type: _graphql.GraphQLString, description: 'the username of the user. Either email or username should be provided' },
    password: { type: (0, _graphql.GraphQLNonNull)(_graphql.GraphQLString) }
  }
});

const login = {
  type: LoginCompletePayload,
  args: {
    input: { type: LoginInput }
  },
  resolve: async (root, args, req) => {
    const user = await (0, _UserAuthentication.logIn)(args.input, req.config, req.auth, req.info && req.info.installationId);
    return (0, _execute.transformResult)('_User', user);
  }
};

function getUserAuthMutationFields() {
  return {
    login
  };
}

exports.default = {
  Query: () => ({}),
  Mutation: getUserAuthMutationFields
};
